lib/cog/template.ex

```diff
@@ -0,0 +1,91 @@
+defmodule Cog.Template do
+  alias Cog.Queries
+  alias Cog.Repo
+  alias Cog.TemplateCache
+
+  @fallback_adapter "any"
+
+  def render(adapter, bundle_id, template, context) do
+    with {:ok, template_fun} <- fetch_compiled_fun(adapter, bundle_id, template, context) do
+      template_fun.(context)
+    end
+  end
+
+  def fetch_compiled_fun(adapter, bundle_id, template, context) do
+    with :error              <- TemplateCache.lookup(adapter, bundle_id, template),
+         {:ok, source}       <- fetch_source(adapter, bundle_id, template, context),
+         {:ok, template_fun} <- compile(source),
+         :ok                 <- TemplateCache.insert(adapter, bundle_id, template, template_fun),
+         do: {:ok, template_fun}
```

Comment:
Would also like feedback here, since this is a little weird with the first match being `:error`.